### PR TITLE
Fix #2109: LocalDate.until sometimes returns wrong result

### LINKED ISSUE #2109: LocalDate.until sometimes returns wrong result
`LocalDate.until(d1, d2)` sometimes returns a wrong result. For instance `LocalDate.of(2013, 3, 1).until(LocalDate.of(2013, 1, 28))` returns a period of -1 months and -1 days, when it should return a period of -1 months and -4 days.

### PR DESCRIPTION
**Pull request by ummels**
Also add private lazy val for proleptic month.

### CODE REVIEW

File: javalib/src/main/scala/java/time/LocalDate.scala [hunk 1]
```diff
@@ -307,17 +310,22 @@ final class LocalDate private (year: Int, month: Month, dayOfMonth: Int)
...
+      if (dmonths > 0 && ddays < 0) -1
+      else if (dmonths < 0 && ddays > 0) 1
+      else 0
+    }
+    val months = dmonths + corr
+    val days = {
+      if (corr < 0) plus(months, MONTHS).until(other, DAYS).toInt
+      else if (corr > 0) ddays - other.lengthOfMonth
+      else ddays
     }
```
Comment (ummels) [inline javalib/src/main/scala/java/time/LocalDate.scala]: This seems to be the only way to stay consistent with the JDK implementation.

**Comment by sjrd** [comment]
Jenkins, test this please

File: javalib/src/main/scala/java/time/LocalDate.scala [hunk 2]
```diff
@@ -31,9 +32,12 @@ final class LocalDate private (year: Int, month: Month, dayOfMonth: Int)
       MathJDK8Bridge.floorDiv(year1, 400).toLong * daysInFourHundredYears
     offset + daysBeforeYear + dayOfYear - 1
   }
+
   private lazy val dayOfWeek =
     MathJDK8Bridge.floorMod(epochDay + 3, 7).toInt + 1
 
+  private lazy val prolepticMonth = year.toLong * 12 + getMonthValue - 1
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/time/LocalDate.scala]: Do we really need to keep this as a field? Wouldn't it be better as a `def`?

**Comment by nicolasstucki** [comment]
In general LGTM. Will just wait on the answer for the field question before merging.

File: javalib/src/main/scala/java/time/LocalDate.scala (hunk 2 above)
Comment (ummels) [inline javalib/src/main/scala/java/time/LocalDate.scala]: Yeah, I guess we can make it a `def` since only a few methods rely on it.

**Comment by ummels** [comment]
Updated. Please retest.

**Comment by nicolasstucki** [comment]
Jenkins, retest this please.

**Pull request merged**